### PR TITLE
Try to fix safeWriteToFile producing empty files on Windows

### PR DESCRIPTION
**Pull request by garymm**
Use win32 APIs to write the temporary file before copying to the final destination. Because we've observed the final file being empty, we suspect that std::ostream::flush is not flushing.

Also add a test for it. It seemed some things in the `fs` namespace were already tested in test_filepath, so I renamed that to test_filesys and added the test there.

## How to test

Unit test tests the happy path. The case where the file rename fails is hard to test, so I think we should just try it and see if the issue is resolved.

Fixes https://github.com/minetest/minetest/issues/9425

### CODE REVIEW

File: src/unittest/test_filesys.cpp [hunk 1]
```diff
@@ -262,3 +265,14 @@ void TestFilePath::testRemoveRelativePathComponent()
...
+
+
+void TestFileSys::testSafeWriteToFile()
+{
+	const std::string dest_path = fs::TempPath() + DIR_DELIM + "testSafeWriteToFile.txt";
+	fs::safeWriteToFile(dest_path, "test");
+	UASSERT(fs::PathExists(dest_path));
+	std::string contents_actual;
+	UASSERT(fs::ReadFile(dest_path, contents_actual));
+	UASSERT(contents_actual == "test");
```
Comment (sfan5) [inline src/unittest/test_filesys.cpp:277]: worth testing that the implementation doesn't accidentally use strlen
```suggestion
	const std::string dest_path = fs::TempPath() + DIR_DELIM + "testSafeWriteToFile.txt";
	constexpr std::string test_data("hello\0world", 11);
	fs::safeWriteToFile(dest_path, test_data);
	UASSERT(fs::PathExists(dest_path));
	std::string contents_actual;
	UASSERT(fs::ReadFile(dest_path, contents_actual));
	UASSERT(contents_actual == test_data);
```

File: src/filesys.cpp [hunk 2]
```diff
@@ -758,14 +758,32 @@ bool safeWriteToFile(const std::string &path, const std::string &content)
 	std::string tmp_file = path + ".~mt";
 
 	// Write to a tmp file
+	bool tmp_success = false;
+
+#ifdef _WIN32
+	// We've observed behavior suggesting that the MSVC implementation of std::ofstream::flush doesn't
+	// actually flush, so we use win32 APIs.
+	HANDLE tmp_handle = CreateFile(
+		tmp_file.c_str(), GENERIC_WRITE, 0, nullptr, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, nullptr);
```
Comment (sfan5) [inline src/filesys.cpp:767]: could maybe use `FILE_ATTRIBUTE_HIDDEN` here too, but dunno if there are wider implications

Comment (garymm) [reply src/filesys.cpp:767]: Seems totally orthogonal to what I'm trying to accomplish here, so I'd rather not in this PR.

File: src/unittest/test_filesys.cpp [hunk 3]
```diff
@@ -262,3 +265,15 @@ void TestFilePath::testRemoveRelativePathComponent()
 	result = fs::RemoveRelativePathComponents(path);
 	UASSERT(result == p("/a/e"));
 }
+
+
+void TestFileSys::testSafeWriteToFile()
+{
+	const std::string dest_path = fs::TempPath() + DIR_DELIM + "testSafeWriteToFile.txt";
+	constexpr std::string test_data("hello\0world", 11);
```
Comment (sfan5) [inline src/unittest/test_filesys.cpp:273]:
```suggestion
	const std::string test_data("hello\0world", 11);
```

**Review by sfan5 — APPROVED** [review]

**Pull request merged**